content/derived-datatypes.rst

```diff
@@ -11,47 +11,461 @@ Derived datatypes
+      Try different combinations of datatypes and adding other fields to the ``struct``.
+
+      Download a :download:`working solution <code/struct-extent-size-solution.c>`
 
-.. todo::
 
-   - Draw diagram showing how to calculate the extent from a typemap.
-   - Type-along showing how to get extent and size of basis datatypes.
+.. typealong:: Extents and the ``count`` parameter
+
+   Let us reiterate: the extent of a custom datatype *is not* its size. The
+   extent tells the MPI runtime how to get to the **next** item in an array of a
+   given type, much like a *stride*.
+
+   We can send an array of ``n`` ``int``-s with a single |term-MPI_Send|:
+
+   .. literalinclude:: code/snippets/send_n.c
+      :language: c
+      :lines: 33-45
+
+   or with ``n`` such calls:
+
+   .. literalinclude:: code/snippets/n_send.c
+      :language: c
+      :lines: 35-46
+
+   In the latter case, we must program explicitly how to get the next element in
+   the array by using the extent of the datatype.
+
+
+Packing and unpacking
+---------------------
+
+MPI offers the possibility to pack and unpack data of known datatype into a
+single contiguous memory buffer, *without* first having to define a
+corresponding datatype.
+This can be an extremely useful technique to reduce messaging traffic and could
+help with the readability and portability of the code.
+The resulting packed buffer will be of type ``MPI_PACKED`` and can contain any
+sort of heterogeneous collection of basic datatypes recognized by MPI.
+
+
+.. figure:: img/E01-pack_unpack.svg
+   :align: center
+
+   MPI allows the programmer to communicate heterogeneous collections into a
+   single message, without defining a full-fledged custom datatype. The data is
+   packed into a buffer of type ``MPI_PACKED``. On the receiving end, the buffer
+   will be unpacked into its constituent components.
+
+
+.. signature:: |term-MPI_Pack|
+
+   Pack data in noncontiguous memory to a contiguous memory buffer.
+
+   .. code-block:: c
+
+      int MPI_Pack(const void *inbuf,
+                   int incount,
+                   MPI_Datatype datatype,
+                   void *outbuf,
+                   int outsize,
+                   int *position,
+                   MPI_Comm comm)
+
+
+.. figure:: img/E01-pack.svg
+   :align: center
+
+   Schematic view of the relation between ``inbuf``, ``outbuf``, and the
+   ``position`` input/output parameter.
+
+
+.. parameters::
+
+   ``inbuf``
+     The input buffer, *i.e.* the data to be packed into contigous memory.
+   ``incount``
+     Number of input data items.
+   ``datatype``
+     The datatype of each item to be packed.
+   ``outbuf``
+     Starting address of the output buffer.
+   ``outsize``
+     The size, in bytes, of the output buffer.
+   ``position``
+     This is an input/output parameter:
+
+     - In *input*, the data in ``inbuf`` will be copied at the address
+       ``outbuf + *position``.
```

Comment:
Did you mean inbuf, rather than outbuf?

Reply:
Double-checked and it's correct as written here. Do you think the following phrasing might be clearer:
```
This is an input/output parameter to keep track of where the current and next, if any, elements should be packed in ``outbuf``:
- In *input*, the data in ``inbuf`` will be copied in ``outbuf`` starting at address ``outbuf + *position``.
- In *output*, it is the first location in ``outbuf`` *after* copying the data: the next element, if any, to be packed will start here. 
```

Reply:
But it is definitely wrong for ``MPI_Unpack``!

Reply:
How about

```
   ``position``
     This is an input/output parameter that describes locations within ``outbuf``. The data at ``inbuf`` will be copied to
     ``outbuf + *position``. After the function returns, the value ``*position`` indicates the first position in ``outbuf``
      that follows the data just copied. (This is useful to pass as ``position`` to the next call to ``MPI_Pack``.)
```

Reply:
:+1: I am redrawing the figure as before-and-after. Together with a better caption it should drive the point home.